spmd/tensor/dispatch.py

```diff
@@ -231,12 +146,107 @@ def operator_dispatch(
             raise NotImplementedError(
                 f"Operator {op_key} does not have a DistributedTensor rule registered."
             )
+        else:
+            return op_schema, False, None
+
+    # step 2. there's sharding propagation rule, run
+    # sharding propagation to get output sharding
+    output_sharding = sharding_prop_func(op_schema)
+
+    # step 3. if can't get output_spec from sharding
+    # propagation (i.e. no rules apply for input
+    # placements), we do auto redistribute on inputs
+    # to get an eligble input, which we will pick a
+    # target schema base on the redistribute cost
+    # TODO: implement full auto distribute with a
+    # simple cost estimation model
+    if output_sharding.output_spec is None:
+        # do auto distributed/boxing here
+        if output_sharding.schema_suggestions is not None:
+            # pick the first suggestion for now,
+            target_schema = output_sharding.schema_suggestions[0]
+            # run sharding propagation again with target schema
+            output_sharding = sharding_prop_func(target_schema)
+
+            return target_schema, True, output_sharding
+
+        else:
+            raise RuntimeError(
+                f"Sharding propagation failed on op {op_key}!"
+                f"Input schema: {op_schema}."
+                f"Failed reason: {output_sharding.failed_reason}"
+            )
+    else:
+        return op_schema, False, output_sharding
+
+
+def operator_dispatch(
+    op_call: torch._ops.OpOverload,
+    args: Tuple[object, ...],
+    kwargs: Dict[str, object],
+    op_to_rules: Dict[str, Callable[[OpSchema], OutputSharding]],
+    custom_dispatch_ops: Dict[str, Callable[..., object]],
+) -> object:
+    # first we need to lift some private aten aliases to public calls
+    if op_call in _CURRENT_DECOMPOSITION_TABLE:
+        return _CURRENT_DECOMPOSITION_TABLE[op_call](*args, **kwargs)
+
+    # STEP 0. See if threre're user defined custom aten operator
+    # implementations. Custom operators take the highest priority
+    if str(op_call) in custom_dispatch_ops:
+        # dispatch to user defined custom distributed tensor ops
+        return custom_dispatch_ops[str(op_call)](*args, **kwargs)
+
+    target_schema, redistribute, output_sharding = prepare_inputs(
+        op_call, args, kwargs, op_to_rules
+    )
+
+    if output_sharding is None:
         # default to local tensor ops, this is wrong
```

Comment:
this comment is invalid as of this change?

Reply:
IIUC, the code below is still calling op on local tensors, no?

Reply:
so if `output_sharding` is None, right now we fall back to executing local tensor ops? I think we disabled this by the `ENABLE_FALLBACK` flag, but now it seems we will still execute this code path?

Reply:
The code snippet below is now in `prepare_inputs`. Basically line 200 will throw if `_ENABLE_FALLBACK=False`. Do you want me to move that check here?

```
        if not _ENABLE_FALLBACK:
            raise NotImplementedError(
                f"Operator {op_key} does not have a DistributedTensor rule registered."
            )
```

Reply:
Ohhh i see, this make sense to me, we don't need to move the checks over, the current impl looks good to me.